go/ct/gen/block_context.go

```diff
@@ -13,43 +13,164 @@
+			if lowerBound < 0 {
+				lowerBound = 0
+			}
+
+			blockNumberRangeSolver.AddLowerBoundary(uint64(lowerBound))
+			blockNumberRangeSolver.AddUpperBoundary(uint64(upperBound))
+
+			// if the range solver is now unsatisfiable, it is most likely because the assigned value is out of the range
+			// of the first number of the current revision - 256.
+			if !blockNumberRangeSolver.IsSatisfiable() {
+				return st.BlockContext{}, ErrUnsatisfiable
+			}
+		}
+	}
+
+	// generate block number
+	blockNumber, err = blockNumberRangeSolver.Generate(rnd)
 	if err != nil {
 		return st.BlockContext{}, err
 	}
+
+	// for all non bound relevante variables, assign them a value based on the constraints.
+	for i, offsetConstraint := range b.variablesOffsetConstraints {
+		if _, isBound := assignment[offsetConstraint.variable]; !isBound {
+			variableRangeSolver := NewRangeSolver[int64](math.MinInt64, math.MaxInt64)
+
+			if difference := offsetConstraint.lowerOffset - offsetConstraint.upperOffset; difference == 0 {
+				// if the difference between the two offsets is 0, then we can only generate a fix value.
+				variableRangeSolver.AddEqualityConstraint(offsetConstraint.lowerOffset)
+			} else if offsetConstraint.lowerOffset > 0 && offsetConstraint.upperOffset < 257 {
+				// if lower offset is greater than 0 and upper is less than 257 we generate IN RANGE
+				variableRangeSolver.AddLowerBoundary(offsetConstraint.upperOffset)
+				variableRangeSolver.AddUpperBoundary(offsetConstraint.lowerOffset)
+			} else {
+				// else have to generate OUT OF RANGE
+				// if blockNumber is too small, then we can ONLY generate OVER the current block number
+				if blockNumber < 256 {
+					variableRangeSolver.AddLowerBoundary(offsetConstraint.upperOffset + int64(blockNumber))
+					variableRangeSolver.AddUpperBoundary(math.MaxInt64)
+
+				} else {
+					// if blockNumber is large enough, we can randomly generate under or over the range.
+					// note that we can only generate under the range if the block number is larger than the upper offset.
+					if rand.Intn(2) == 0 && blockNumber-uint64(offsetConstraint.upperOffset) >= blockNumber {
+						// generate under the range
+						variableRangeSolver.AddLowerBoundary(0)
+						variableRangeSolver.AddUpperBoundary(int64(blockNumber))
+
+					} else {
+						// generate over the range
+						variableRangeSolver.AddLowerBoundary(math.MinInt64)
+						variableRangeSolver.AddUpperBoundary(0)
+					}
+				}
+			}
+			newValue, err := variableRangeSolver.Generate(rnd)
+			if err != nil {
+				return st.BlockContext{}, err
+			}
+			assignment[offsetConstraint.variable] = common.NewU256(uint64(int64(blockNumber) - newValue))
+		} else {
+			// if the variable is bound, then we need to check that it does not clash with any other constraint.
+			for _, previousConstraint := range b.variablesOffsetConstraints[:i] {
+				if previousConstraint.variable == offsetConstraint.variable {
+
```

Comment:
Can you explain to me why is it not enough with testing that the assigned value is in or out of the constraint range? 
as I understand, both B and the offset variable ($a) have been assigned at this point, and the only reason why it would not be satisfiable is if $a is not in the constraint, right?

Reply:
discusses offline. We agreed on a simpler way to check this.